File: 143.reorder-list.java
import javax.management.ListenerNotFoundException;

/*
 * @lc app=leetcode id=143 lang=java
 *
 * [143] Reorder List
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public void reorderList(ListNode head) {
        if (head == null || head.next == null) {
            return;
        }

        ListNode slow = head;
        ListNode fast = head;
        
        // divide the list into 2 sbu-lists
        while (fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;
        }

        // remember we need to wet slow to be null 
        // which means cut off from the second list
        ListNode mid = slow.next;
        slow.next = null;

        ListNode subHead = rotateList(mid);

        // merge two lists
        while (head != null && subHead != null) {
            ListNode temp = head.next;
            head.next = subHead;
            subHead = subHead.next;
            head.next.next = temp;
            head = temp;
        }
    }

    public ListNode rotateList(ListNode head) {
        ListNode newHead = null;

        while (head != null) {
            ListNode next = head.next;
            head.next = newHead;
            newHead = head;
            head = next;
        }

        return newHead;
    }
}
// @lc code=end

File: 143.reorder-list.cpp
/*
 * @lc app=leetcode id=143 lang=cpp
 *
 * [143] Reorder List
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    void reorderList(ListNode* head) {
        ListNode* dummy = new ListNode(-1);
        dummy->next = head;

        ListNode* fast = dummy;
        ListNode* slow = dummy;
        while (fast && fast->next) {
            slow = slow->next;
            fast = fast->next->next;
        }

        ListNode* tail = reverse(slow->next);
        slow->next = NULL;


        ListNode* cur = dummy->next;
        while (cur && tail) {
            ListNode* tmp1 = cur->next;
            ListNode* tmp2 = tail->next;

            cur->next = tail;
            tail->next = tmp1;

            cur = tmp1;
            tail = tmp2;
        }

        head = dummy->next;
    }

    ListNode* reverse(ListNode* head) {
        if (!head || !head->next) {
            return head;
        }

        ListNode* newHead = reverse(head->next);
        head->next->next = head;
        head->next = NULL; 
        
        return newHead;
    }
};
// @lc code=end

